fix: add Byline and Selection features to List component

InnaAtanasova opened the pull request:
## Related Issue
Closes SAP/fundamental-styles#853
Closes SAP/fundamental-styles#865
## Description
- This PR introduces the List With Byline functionality which was missing in Standard List component.
- It also introduces the Selection functionality
- Added rules for the empty radio/checkbox label
- The current PR introduces no Breaking Changes for the List component.

InnaAtanasova commented on src/list.scss:
```diff
@@ -387,4 +403,138 @@ $block: #{$fd-namespace}-list;
+      @include fd-reset();
+      @include fd-flex-center();
+
+      width: $fd-list-byline-thumbnail-dimensions;
+      height: $fd-list-byline-thumbnail-dimensions;
+      min-width: $fd-list-byline-thumbnail-dimensions;
+      max-width: $fd-list-byline-thumbnail-dimensions;
+      min-height: $fd-list-byline-thumbnail-dimensions;
+      max-height: $fd-list-byline-thumbnail-dimensions;
```
If you wonder why I have to use min-width, width, max-width and set them to the same value it's because we need to handle the case when an image (fd-image) is used and for this reason it needs to overwrite the min-width/max-width, min-height/max-height properties of the image. We are currently migrating Image to Avatar so this might be refactored later to only use width and height.

JKMarkowski replied:
It also helps IE11 to interpret the code. Sometimes it's even mandatory to keep fixed width.

rengare commented on src/list.scss:
```diff
@@ -387,4 +427,158 @@ $block: #{$fd-namespace}-list;
+    }
+
+    .#{$block}__form-item {
+      margin-left: -1rem;
```
value can be moved to variable since it can be reused
**Edit** there are few places when values can be moved to variables

InnaAtanasova replied:
I will go over the values that are used more than once and put them in variables.
Edit2 comment: this is not a List with Navigation Indicator. This is just an example that the icon (ANY icon) can be placed on the right side. I can use another icon in the documentation

InnaAtanasova commented on src/mixins/_forms.scss:
```diff
@@ -155,6 +155,21 @@
   }
 }
 
+@mixin fd-form-radio-empty-focus($margin) {
```
![Screen Shot 2020-04-20 at 10 32 54 AM](https://user-images.githubusercontent.com/39598672/79763709-6bc52480-82f2-11ea-96d3-d8adae9340c1.png)
![Screen Shot 2020-04-20 at 10 32 40 AM](https://user-images.githubusercontent.com/39598672/79763714-6c5dbb00-82f2-11ea-8ca4-61e5be3e8585.png)

This is how checkbox looks like without this mixin. You can open an issue for this and we can work on changing the implementation of the checkbox.

InnaAtanasova commented on src/list.scss:
```diff
@@ -387,4 +427,158 @@ $block: #{$fd-namespace}-list;
+        .#{$block}__form-item {
+          margin-top: $fd-list-form-item-margin-top;
+
+          label {
```
It could be a checkbox label (`fd-checkbox__label`) or a radiobox label (`fd-radio__label`) or any other checkbox/radio so I'm sure which is better

InnaAtanasova commented on src/list.scss:
```diff
@@ -387,4 +427,158 @@ $block: #{$fd-namespace}-list;
+      color: var(--sapContent_LabelColor);
+      line-height: $fd-list-byline-text-line-height;
+
+      &--2-col {
```
I disagree. The sections should be 2 and by not specifying this the user can think that multiple sections are allowed.

InnaAtanasova commented on src/list.scss:
```diff
@@ -164,6 +174,36 @@ $block: #{$fd-namespace}-list;
+    padding: 0 $fd-list-horizontal-padding;
+    cursor: pointer;
+
+    @include fake-list-outline();
```
We need an offset for the outline and the offset property is not supported in IE11. This solution was discussed few weeks ago and we agreed to use it (if possible, because it relies on ::before). Any other solutions are more than welcome.

InnaAtanasova commented on src/list.scss:
```diff
@@ -103,15 +106,22 @@ $block: #{$fd-namespace}-list;
     }
   }
 
+  &__form-item {
```
The base `fd-form-item` does not have properties margin-top and margin-left that `fd-list__form-item` is overwriting so there's no chance of conflict where we will depend on the order of import. This is an issue when the prop exists in the base class and we overwrite it, like the case with the background color. But anyways, doesn't hurt to change it so I did it :)

InnaAtanasova replied:
It now has both: 
```
.fd-list__form-item,
.fd-list__form-item.fd-list-item  {...}
```

since we need to test the selection and we cannot use `fd-list-item` inside the List test.

InnaAtanasova commented on src/list.scss:
```diff
@@ -387,4 +427,158 @@ $block: #{$fd-namespace}-list;
+      max-height: $fd-list-byline-thumbnail-dimensions;
+      border-radius: $fd-list-byline-thumbnail-border-radius;
+      margin-right: $fd-list-byline-thumbnail-spacing;
+      font-size: 2.5rem;
```
Not used anywhere else

InnaAtanasova commented:
> * I think we are missing here this element (it is a part of selection list):
>   [List (Fiori 3) - Selection - Selected Items Indicator Bar](https://wiki.wdf.sap.corp/wiki/display/visualcore/List+%28Fiori+3%29+-+Selection+-+Selected+Items+Indicator+Bar)

It's missing because it depends on Toolbar which we don't have yet. I need to work on Navigation Indication which needs a List Item with selection so I can later start the Message View. I created an issue for this: https://github.com/SAP/fundamental-styles/issues/880

InnaAtanasova commented:
> * Is empty secondary text a valid test?
>   ![image](https://user-images.githubusercontent.com/17496353/79738808-9072c980-82fd-11ea-9d28-5a4bb5af82f8.png)

<img width="937" alt="Screen Shot 2020-04-20 at 12 34 11 PM" src="https://user-images.githubusercontent.com/39598672/79776147-4ab8ff80-8303-11ea-9f5b-5cafb9313d42.png">
 I am not the person to say coz I didn't introduce this but I think it's intentional based on the screenshot. I think it's to show a case where a list can have items with and without secondary value.

InnaAtanasova commented on src/list.scss:
```diff
@@ -387,4 +427,158 @@ $block: #{$fd-namespace}-list;
+      min-height: $fd-list-byline-thumbnail-dimensions;
+      height: 100%;
+      width: 100%;
+      overflow: hidden;
```
Removed it

InnaAtanasova commented:
> * There are no tests for list with checkbox
There wasn't much sense of having tests for this but I added some 

> * This PR is missing `aria.role` attributes for list and list with select
Added

JKMarkowski commented on src/checkbox.scss:
```diff
@@ -83,6 +83,21 @@ $block: #{$fd-namespace}-checkbox;
+      outline: none;
+    }
+
+    &:empty {
+      padding: $fd-checkbox-margin;
+      margin: 0;
+
+      &::before,
+      &::after {
+        margin-right: 0;
+        margin-left: 0;
+      }
+    }
```
Will `&:empty` work on angular, which adds some comments to template ?

mikerodonnell89 replied:
Just checked, the `&:empty` psuedo selector does not get applied to the angular checkbox

JKMarkowski commented on src/list.scss:
```diff
@@ -387,4 +433,161 @@ $block: #{$fd-namespace}-list;
+    .#{$block}__form-item,
+    .#{$block}__form-item.#{$form-item} {
+      margin-left: $fd-list-form-item-margin-x;
+      margin-top: -0.6875rem;
```
Crazy, how complicated it has to be, to keep focus on IE11
